src/routing/bootstrap.rs

```diff
@@ -336,70 +173,125 @@ impl<'a> State<'a> {
+                    return Err(Error::TryJoinLater);
+                }
                 JoinResponse::Approval {
                     section_auth,
                     genesis_key,
                     section_chain,
+                    ..
                 } => {
                     return Ok((
                         self.node,
                         Section::new(genesis_key, section_chain, section_auth)?,
                         self.backlog.into_iter().collect(),
                     ));
                 }
-                JoinResponse::Retry {
-                    section_auth,
-                    section_key: new_section_key,
-                } => {
-                    if new_section_key == section_key {
+                JoinResponse::Retry(section_auth) | JoinResponse::Redirect(section_auth) => {
+                    if section_auth.section_key() == section_key {
                         continue;
                     }
 
-                    if section_auth.prefix().matches(&self.node.name()) {
+                    // Ignore already used recipients
+                    let new_recipients: Vec<(XorName, SocketAddr)> = section_auth
+                        .elders
+                        .iter()
+                        .filter(|(name, addr)| !used_recipient.contains(&(**name, **addr)))
+                        .map(|(name, addr)| (*name, *addr))
+                        .collect();
+
+                    if new_recipients.is_empty() {
+                        debug!("Joining redirected to the same set of peers we already contacted - ignoring response");
+                    } else {
+                        info!(
+                            "Joining redirected to another set of peers: {:?}",
+                            new_recipients,
+                        );
+                    }
+
+                    let prefix = section_auth.prefix;
+
+                    // For the first section, using age random among 6 to 100 to avoid
+                    // relocating too many nodes at the same time.
+                    if prefix.is_empty() && self.node.age() < FIRST_SECTION_MIN_AGE {
+                        let age: u8 = (FIRST_SECTION_MIN_AGE..FIRST_SECTION_MAX_AGE)
+                            .choose(&mut rand::thread_rng())
+                            .unwrap_or(FIRST_SECTION_MAX_AGE);
+
+                        let new_keypair =
+                            crypto::gen_keypair(&Prefix::default().range_inclusive(), age);
+                        let new_name = crypto::name(&new_keypair.public);
+
+                        info!("Setting Node name to {}", new_name);
+                        self.node = Node::new(new_keypair, self.node.addr);
+                    }
+
+                    // if we are relocating, and we didn't generate
+                    // the relocation payload yet, we do it now
+                    if relocate_payload.is_none() {
```

Comment:
This `relocate_payload` is just be set to None at the beginning of the loop .
or does the flow changed to sending back `relocate_payload` via the JoinResponse ?

Reply:
Yes, but it's outside of the loop. 
So the idea is that the first request will won't send the relocate payload since we don't know the section prefix (similar to what it used to happen when we would be first getting info from the section with a GetSectionInfo request, but now we send a JoinRequest to grab that same info), once we get the first JoinResponse::Retry (which contains the prefix) we will populate it with the relocate payload.

Reply:
maybe the boostrap flow got changed, however the relocate flow as I remember is:

Reply:
yes, but previously we still send a GetSectionInfo to grab the prefix from the target section, so here we do the same

Reply:
yes, but previously `relocate_details` is a passed_in populated value, before the GetSectionInfo call.
now it's set to None before the call.